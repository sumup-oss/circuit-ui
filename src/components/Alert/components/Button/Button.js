import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';

const CENTER = 'center';
const TOP = 'top';

const baseStyles = ({ theme }) => css`
  label: alert__button;
  display: block;
  padding-left: ${theme.spacings.giga};
  margin-left: auto;
  flex-grow: 0;
  flex-shrink: 0;
`;

const alignmentStyles = ({ alignment }) => {
  const alignments = {
    [CENTER]: 'center',
    [TOP]: 'flex-start'
  };
  return css`
    label: alert__button--${alignment};
    align-self: ${alignments[alignment]};
  `;
};

/**
 * Button used in the Alert component. Used for styling and aligment
 * pruposes only.
 */
const AlertButton = styled('div')(baseStyles, alignmentStyles);

AlertButton.TOP = TOP;
AlertButton.CENTER = CENTER;

AlertButton.propTypes = {
  /**
   * Button
   */
  children: PropTypes.element.isRequired,
  /**
   * Vertical alignment
   */
  alignment: PropTypes.oneOf([AlertButton.TOP, AlertButton.CENTER])
};

AlertButton.defaultProps = {
  alignment: AlertButton.CENTER
};

/**
 * @component
 */
export default AlertButton;
