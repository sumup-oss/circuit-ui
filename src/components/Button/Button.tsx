/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { HTMLProps, ReactNode, ReactElement } from 'react';
import { css } from '@emotion/core';

import styled, { StyleProps } from '../../styles/styled';
import {
  textMega,
  disableVisually,
  focusOutline
} from '../../styles/style-helpers';
import { useComponents } from '../ComponentsContext';

export interface BaseProps {
  children: ReactNode;
  /**
   * Choose from 3 style variants. Default: 'primary'.
   */
  variant?: 'primary' | 'secondary' | 'tertiary';
  /**
   * Choose from 3 sizes. Default: 'mega'.
   */
  size?: 'kilo' | 'mega' | 'giga';
  /**
   * Visually and functionally disable the button.
   */
  disabled?: boolean;
  /**
   * Stretch the button across the full width of its parent.
   */
  stretch?: boolean;
}

type LinkElProps = Omit<HTMLProps<HTMLAnchorElement>, 'size' | 'type'>;
type ButtonElProps = Omit<HTMLProps<HTMLButtonElement>, 'size' | 'type'>;

export type ButtonProps = BaseProps & LinkElProps & ButtonElProps;

type ReturnType = ReactElement<any, any> | null;

const BORDER_WIDTH = '1px';

const baseStyles = ({ theme }: StyleProps) => css`
  label: button;
  display: flex;
  align-items: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  ${textMega({ theme })};
  text-align: center;
  text-decoration: none;
  font-weight: ${theme.fontWeight.bold};
  border-radius: ${theme.borderRadius.giga};
  border-width: ${BORDER_WIDTH};
  border-style: solid;

  &:focus {
    ${focusOutline({ theme })};
  }

  &:disabled,
  &[disabled] {
    ${disableVisually()};
  }
`;

const primaryStyles = ({
  theme,
  variant = 'primary'
}: ButtonProps & StyleProps) =>
  variant === 'primary' &&
  css`
    label: button--primary;
    background-color: ${theme.colors.p500};
    border-color: ${theme.colors.p500};
    color: ${theme.colors.white};

    &:hover {
      background-color: ${theme.colors.p700};
      border-color: ${theme.colors.p700};
    }

    &:active {
      background-color: ${theme.colors.p900};
      border-color: ${theme.colors.p900};
    }
  `;

const secondaryStyles = ({
  theme,
  variant = 'primary'
}: ButtonProps & StyleProps) =>
  variant === 'secondary' &&
  css`
    label: button--secondary;
    background-color: ${theme.colors.white};
    border-color: ${theme.colors.n500};
    color: ${theme.colors.n800};

    &:hover {
      background-color: ${theme.colors.n100};
      border-color: ${theme.colors.n700};
    }

    &:active {
      background-color: ${theme.colors.n200};
      border-color: ${theme.colors.n700};
    }
  `;

const tertiaryStyles = ({
  theme,
  variant = 'primary'
}: ButtonProps & StyleProps) =>
  variant === 'tertiary' &&
  css`
    label: button--tertiary;
    background-color: transparent;
    border-color: transparent;
    color: ${theme.colors.p500};

    &:hover {
      color: ${theme.colors.p700};
    }

    &:active {
      color: ${theme.colors.p900};
    }
  `;

const sizeStyles = ({ theme, size = 'mega' }: ButtonProps & StyleProps) => {
  const sizeMap = {
    kilo: `${theme.spacings.bit} calc(${theme.spacings.mega} - ${BORDER_WIDTH})`,
    mega: `${theme.spacings.byte} calc(${theme.spacings.giga} - ${BORDER_WIDTH})`,
    giga: `${theme.spacings.kilo} calc(${theme.spacings.tera} - ${BORDER_WIDTH})`
  };

  return css({
    label: `button--${size}`,
    padding: sizeMap[size]
  });
};

const stretchStyles = ({ stretch }: ButtonProps) =>
  stretch &&
  css`
    label: button--stretch;
    width: 100%;
  `;

const BaseButton = styled('button')<ButtonProps>(
  baseStyles,
  primaryStyles,
  secondaryStyles,
  tertiaryStyles,
  sizeStyles,
  stretchStyles
);

/**
 * The Button component enables the user to perform an action or navigate
 * to a different screen.
 */
export function Button(props: BaseProps & LinkElProps): ReturnType;
export function Button(props: BaseProps & ButtonElProps): ReturnType;
export function Button(props: ButtonProps): ReturnType {
  const { Link } = useComponents();
  const LinkButton = BaseButton.withComponent(Link);
  return props.href ? <LinkButton {...props} /> : <BaseButton {...props} />;
}
