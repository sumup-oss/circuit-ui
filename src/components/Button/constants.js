import PropTypes from 'prop-types';

import { sizes } from '../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

export const LOADING_STATES = {
  INACTIVE: 'inactive',
  ACTIVE: 'active',
  SUCCESS: 'success',
  ERROR: 'error'
};

export const EXIT_ANIMATION_DURATION = 2000;

export const SIZE_PROP_TYPE = PropTypes.oneOf([KILO, MEGA, GIGA]);
