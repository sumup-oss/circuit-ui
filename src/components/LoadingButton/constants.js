export const EXIT_ANIMATION_DURATION = 2000;

export const DEFAULT_LOADING_DELAY = 500;

export const LOADING_STATES = {
  DISABLED: 'disabled',
  ACTIVE: 'active',
  SUCCESS: 'success',
  ERROR: 'error'
};
