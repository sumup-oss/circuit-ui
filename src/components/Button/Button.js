/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { omit } from 'lodash/fp';
import PropTypes from 'prop-types';

import PlainButton from './components/PlainButton';
import RegularButton from './components/RegularButton';
import { sizes } from '../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

const SHARED_PROPS = [
  'children',
  'className',
  'data-selector',
  'disabled',
  'href',
  'onClick',
  'primary',
  'size',
  'type'
];

export const BUTTON_PROPS = [
  ...SHARED_PROPS,
  'deepRef',
  'as',
  'secondary',
  'size',
  'stretch',
  'target'
];

export const PLAIN_BUTTON_PROPS = [...SHARED_PROPS, 'size', 'target'];

const REGULAR_BUTTON_ONLY_PROPS = ['deepRef', 'as', 'secondary', 'stretch'];

/**
 * A button component with support for the anchor and button
 * element as well as a button-looking button and a text link.
 */
const Button = ({ plain, ...props }) =>
  plain ? (
    <PlainButton {...omit(REGULAR_BUTTON_ONLY_PROPS, props)} />
  ) : (
    <RegularButton {...props} />
  );

Button.KILO = KILO;
Button.MEGA = MEGA;
Button.GIGA = GIGA;

Button.propTypes = {
  /**
   * Should the Button be disabled?
   */
  disabled: PropTypes.bool,
  /**
   * URL the Button should lead to. Causes the Button to render an <a> tag.
   */
  href: PropTypes.string,
  /**
   * Makes the button look and behave like a text link.
   */
  plain: PropTypes.bool,
  /**
   * Renders a primary button using the brand color.
   */
  primary: PropTypes.bool,
  /**
   * Size of the button. Use the Button's KILO, MEGA, or GIGA properties.
   */
  size: PropTypes.oneOf([Button.KILO, Button.MEGA, Button.GIGA]),
  /**
   * Renders a secondary button.
   */
  secondary: PropTypes.bool,
  /**
   * Trigger stretch (full width) styles on the component.
   */
  stretch: PropTypes.bool,
  /**
   * Link target. Should only be passed, if href is passed, too.
   */
  target: PropTypes.string
};

Button.defaultProps = {
  disabled: false,
  href: null,
  plain: false,
  primary: false,
  secondary: false,
  size: Button.MEGA,
  stretch: false,
  target: null
};

/**
 * @component
 */
export default Button;
