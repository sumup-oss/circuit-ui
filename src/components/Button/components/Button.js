import React from 'react';
import styled, { css } from 'react-emotion';
import PropTypes from 'prop-types';

import { textMega } from '../../../styles/style-helpers';
import { sizes } from '../../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

const calculatePadding = ({ theme, size: buttonSize }) => (diff = '0px') => {
  const sizeMap = {
    [KILO]: `calc(${theme.spacings.bit} - ${diff}) calc(${
      theme.spacings.mega
    } - ${diff})`,
    [MEGA]: `calc(${theme.spacings.byte} - ${diff}) calc(${
      theme.spacings.giga
    } - ${diff})`,
    [GIGA]: `calc(${theme.spacings.kilo} - ${diff}) calc(${
      theme.spacings.tera
    } - ${diff})`
  };

  if (!sizeMap[buttonSize] && buttonSize) {
    return null;
  }

  return sizeMap[buttonSize] || sizeMap.mega;
};

const disabledStyles = css`
  label: button--disabled;
  opacity: 0.4;
  pointer-events: none;
`;

const stretchStyles = ({ stretch }) =>
  stretch &&
  css`
    label: button--stretched;
    width: 100%;
    display: block;
  `;

const baseStyles = ({ theme, href, ...otherProps }) => css`
  label: button;
  background-color: ${theme.colors.n100};
  border-color: ${theme.colors.n300};
  border-radius: ${theme.borderRadius.mega};
  border-style: solid;
  border-width: 1px;
  box-shadow: inset 0 1px 0 1px rgba(255, 255, 255, 0.06);
  display: ${href ? 'inline-block' : 'block'};
  color: ${theme.colors.n700};
  cursor: pointer;
  font-weight: ${theme.fontWeight.bold};
  width: auto;
  height: auto;
  text-decoration: none;
  ${textMega({ theme })};

  &:active {
    background-color: ${theme.colors.n300};
    border-color: ${theme.colors.n500};
    box-shadow: inset 0 4px 8px 0 rgba(12, 15, 20, 0.3);
  }

  &:focus {
    border-color: ${theme.colors.n500};
    border-width: 2px;
    outline: 0;
    padding: ${calculatePadding({ theme, ...otherProps })('1px')};
  }

  &:hover {
    background-color: ${theme.colors.n300};
  }

  &:hover,
  &:active {
    border-color: ${theme.colors.n500};
    border-width: 1px;
    padding: ${calculatePadding({ theme, ...otherProps })()};
  }

  &[disabled],
  &:disabled {
    ${disabledStyles};
  }
`;

const primaryStyles = ({ theme, primary }) =>
  primary &&
  css`
    label: button--primary;
    background-color: ${theme.colors.p500};
    border-color: ${theme.colors.p700};
    color: ${theme.colors.white};

    &:active {
      background-color: ${theme.colors.p700};
      border-color: ${theme.colors.p900};
    }

    &:focus {
      border-color: ${theme.colors.p700};
    }

    &:hover {
      background-color: ${theme.colors.p700};
    }

    &:hover,
    &:active {
      border-color: ${theme.colors.p900};
    }
  `;

const flatStyles = ({ theme, flat, secondary, ...otherProps }) =>
  flat &&
  !secondary &&
  css`
    label: button--flat;
    border-width: 0px;
    box-shadow: 0 0 0 1px rgba(12, 15, 20, 0.02),
      0 2px 2px 0 rgba(12, 15, 20, 0.06), 0 4px 4px 0 rgba(12, 15, 20, 0.06);

    &:active {
      background-color: ${theme.colors.p900};
      box-shadow: 0 0 0 1px rgba(12, 15, 20, 0.02),
        0 0 1px 0 rgba(12, 15, 20, 0.06), 0 2px 2px 0 rgba(12, 15, 20, 0.06);
    }

    &:active:focus,
    &:hover:focus {
      border-width: 0;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })()};
    }

    &:focus {
      border-width: 2px;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })(
        '2px'
      )};
    }
  `;

const secondaryStyles = ({ theme, secondary, flat, ...otherProps }) =>
  secondary &&
  css`
    label: button--secondary;
    background-color: transparent;
    border-color: ${theme.colors.n900};
    border-width: 0;
    box-shadow: none;
    color: ${theme.colors.n700};

    &:active {
      background-color: transparent;
      border-color: ${theme.colors.n900};
      border-width: 0;
      box-shadow: none;
    }

    &:focus {
      border-color: ${theme.colors.n900};
      border-width: 2px;
      box-shadow: none;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })(
        '2px'
      )};
    }

    &:hover {
      background-color: transparent;
      border-width: 0;
      border-color: ${theme.colors.n900};
    }

    &:active,
    &:hover,
    &:hover:focus,
    &:active:focus {
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })()};
    }

    &:active,
    &:hover,
    &:focus {
      color: ${theme.colors.n900};
    }

    &:active:focus,
    &:hover:focus {
      border-color: ${theme.colors.n900};
      border-width: 2px;
      box-shadow: none;
      padding: ${calculatePadding({ theme, flat, secondary, ...otherProps })(
        '2px'
      )};
    }
  `;

const sizeStyles = props => {
  const { size: buttonSize } = props;
  const padding = calculatePadding(props)();
  return css({
    label: `button--${buttonSize}`,
    padding
  });
};

const buttonLoadingStyles = ({ isLoading }) =>
  isLoading &&
  css`
    label: button--loading;
    pointer-events: none;
  `;

const ButtonElement = styled('button')`
  ${baseStyles};
  ${primaryStyles};
  ${sizeStyles};
  ${flatStyles};
  ${secondaryStyles};
  ${stretchStyles};
  ${buttonLoadingStyles};
`;

const LinkButtonElement = ButtonElement.withComponent('a');

const Button = ({ href, ...props }) =>
  href ? (
    <LinkButtonElement {...{ ...props, href }} />
  ) : (
    <ButtonElement {...props} />
  );

Button.propTypes = {
  /**
   * URL the Button should lead to. Causes the Button to render an <a> tag.
   */
  href: PropTypes.string,
  /**
   * Should the Button be disabled?
   */
  disabled: PropTypes.bool,
  /**
   * Button has a 'flat' variation, triggered with this prop.
   */
  flat: PropTypes.bool,
  /**
   * Renders a secondary button. Secondary buttons look the same for
   * primary (default) and flat buttons.
   */
  secondary: PropTypes.bool,
  /**
   * Renders a primary button using the brand color.
   */
  primary: PropTypes.bool,
  /**
   * Link target. Should only be passed, if href is passed, too.
   */
  target: PropTypes.string,
  /**
   * Size of the button. Use the Button's KILO, MEGA, or GIGA properties.
   */
  size: PropTypes.oneOf([Button.KILO, Button.MEGA, Button.GIGA]),
  /**
   * Trigger stretch (full width) styles on the component.
   */
  stretch: PropTypes.bool
};

Button.defaultProps = {
  disabled: false,
  flat: false,
  size: MEGA,
  target: null,
  href: null,
  primary: false,
  secondary: false,
  stretch: false
};

/**
 * @component
 */
export default Button;
