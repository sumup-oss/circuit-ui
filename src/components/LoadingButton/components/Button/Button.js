/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import Button from '../../../Button';
import LoadingIcon from '../LoadingIcon';
import { DISABLED, ACTIVE, SUCCESS, ERROR } from '../../constants';
import {
  BUTTON_PROP_TYPES,
  BUTTON_DEFAULT_PROPS
} from '../../../Button/constants';

const childrenWrapperStyles = ({ theme }) => css`
  opacity: 1;
  transform: translate(0, 0);
  transition: opacity ${theme.transitions.default},
    transform ${theme.transitions.default};
`;

const childrenWrapperLoadingStyles = ({ loadingState }) =>
  loadingState !== DISABLED &&
  css`
    opacity: 0;
    transform: translate(0, 100%);
  `;

const Wrapper = styled.div`
  position: relative;
`;

const ChildrenWrapper = styled.div`
  ${childrenWrapperStyles};
  ${childrenWrapperLoadingStyles};
`;

const LoadingButton = ({
  loadingState,
  size,
  children,
  onClick,
  isLoading,
  ...otherProps
}) => (
  <Button
    {...otherProps}
    size={size}
    onClick={isLoading ? null : onClick}
    isLoading={isLoading}
  >
    <Wrapper>
      <LoadingIcon size={size} loadingState={loadingState} />
      <ChildrenWrapper loadingState={loadingState}>{children}</ChildrenWrapper>
    </Wrapper>
  </Button>
);

LoadingButton.propTypes = {
  ...BUTTON_PROP_TYPES,
  loadingState: PropTypes.oneOf([DISABLED, ACTIVE, SUCCESS, ERROR]),
  isLoading: PropTypes.bool
};

LoadingButton.defaultProps = {
  ...BUTTON_DEFAULT_PROPS,
  loadingState: DISABLED,
  isLoading: false
};

export default LoadingButton;
