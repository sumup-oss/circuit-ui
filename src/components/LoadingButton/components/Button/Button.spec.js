/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import { ACTIVE } from '../../constants';
import Button from '.';

describe('Button', () => {
  describe('Style tests', () => {
    it('should have default styles', () => {
      const actual = create(<Button />);
      expect(actual).toMatchSnapshot();
    });

    it('should have active loading styles', () => {
      const actual = create(<Button loadingState={ACTIVE} />);
      expect(actual).toMatchSnapshot();
    });

    it('should have isLoading styles', () => {
      const actual = create(<Button isLoading />);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('Logic tests', () => {
    it('should not bind the click handler to the button while loading', () => {
      const onClick = jest.fn();
      const { getByTestId } = render(
        <Button isLoading onClick={onClick} data-testid="loading-button" />
      );

      act(() => {
        fireEvent.click(getByTestId('loading-button'));
      });

      expect(onClick).not.toHaveBeenCalled();
    });
  });

  describe('Accessibility tests', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<Button>Loading Button</Button>);
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
