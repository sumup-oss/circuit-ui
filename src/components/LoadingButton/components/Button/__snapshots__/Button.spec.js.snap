// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Style tests should have active loading styles 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate3d(0,0,1,0deg);
    -ms-transform: rotate3d(0,0,1,0deg);
    transform: rotate3d(0,0,1,0deg);
  }

  100% {
    -webkit-transform: rotate3d(0,0,1,360deg);
    -ms-transform: rotate3d(0,0,1,360deg);
    transform: rotate3d(0,0,1,360deg);
  }
}

@keyframes animation-0 {
  0% {
    -webkit-transform: rotate3d(0,0,1,0deg);
    -ms-transform: rotate3d(0,0,1,0deg);
    transform: rotate3d(0,0,1,0deg);
  }

  100% {
    -webkit-transform: rotate3d(0,0,1,360deg);
    -ms-transform: rotate3d(0,0,1,360deg);
    transform: rotate3d(0,0,1,360deg);
  }
}

.circuit-7 {
  background-color: #FAFBFC;
  border-color: #D8DDE1;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: inset 0 1px 0 1px rgba(255,255,255,0.06);
  display: block;
  color: #5C656F;
  cursor: pointer;
  font-weight: 700;
  width: auto;
  height: auto;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  line-height: 24px;
  padding: calc(8px - 0px) calc(24px - 0px);
  position: relative;
  overflow: hidden;
}

.circuit-7:active {
  background-color: #D8DDE1;
  border-color: #9DA7B1;
  box-shadow: inset 0 4px 8px 0 rgba(12,15,20,0.3);
}

.circuit-7:focus {
  border-color: #9DA7B1;
  border-width: 2px;
  outline: 0;
  padding: calc(8px - 1px) calc(24px - 1px);
}

.circuit-7:hover {
  background-color: #D8DDE1;
}

.circuit-7:hover,
.circuit-7:active {
  border-color: #9DA7B1;
  border-width: 1px;
  padding: calc(8px - 0px) calc(24px - 0px);
}

.circuit-7[href] {
  display: inline-block;
}

.circuit-7:disabled,
.circuit-7[disabled] {
  opacity: 0.4;
  pointer-events: none;
  -webkit-user-selectable: none;
  -moz-user-selectable: none;
  -ms-user-selectable: none;
  user-selectable: none;
}

.circuit-2 {
  height: 24px;
  width: 24px;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.circuit-0 {
  width: 24px;
  height: 24px;
  border-radius: 100%;
  border: 2px solid #FFFFFF;
  border-top-color: transparent;
  -webkit-animation: animation-0 1s infinite linear;
  animation: animation-0 1s infinite linear;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  opacity: 0;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  opacity: 1;
}

.circuit-4 {
  display: inline-block;
  opacity: 1;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  -webkit-transition: opacity 200ms ease-in-out,-webkit-transform 200ms ease-in-out;
  -webkit-transition: opacity 200ms ease-in-out,transform 200ms ease-in-out;
  transition: opacity 200ms ease-in-out,transform 200ms ease-in-out;
  opacity: 0;
  -webkit-transform: translate(0,100%);
  -ms-transform: translate(0,100%);
  transform: translate(0,100%);
}

<button
  class="circuit-6 circuit-7 circuit-8"
>
  <div
    class="circuit-2 circuit-3"
  >
    <div
      aria-busy="true"
      aria-live="assertive"
      class="circuit-0 circuit-1"
      role="alert"
    />
  </div>
  <span
    class="circuit-4 circuit-5"
  />
</button>
`;

exports[`Button Style tests should have default styles 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate3d(0,0,1,0deg);
    -ms-transform: rotate3d(0,0,1,0deg);
    transform: rotate3d(0,0,1,0deg);
  }

  100% {
    -webkit-transform: rotate3d(0,0,1,360deg);
    -ms-transform: rotate3d(0,0,1,360deg);
    transform: rotate3d(0,0,1,360deg);
  }
}

.circuit-7 {
  background-color: #FAFBFC;
  border-color: #D8DDE1;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: inset 0 1px 0 1px rgba(255,255,255,0.06);
  display: block;
  color: #5C656F;
  cursor: pointer;
  font-weight: 700;
  width: auto;
  height: auto;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  line-height: 24px;
  padding: calc(8px - 0px) calc(24px - 0px);
  position: relative;
  overflow: hidden;
}

.circuit-7:active {
  background-color: #D8DDE1;
  border-color: #9DA7B1;
  box-shadow: inset 0 4px 8px 0 rgba(12,15,20,0.3);
}

.circuit-7:focus {
  border-color: #9DA7B1;
  border-width: 2px;
  outline: 0;
  padding: calc(8px - 1px) calc(24px - 1px);
}

.circuit-7:hover {
  background-color: #D8DDE1;
}

.circuit-7:hover,
.circuit-7:active {
  border-color: #9DA7B1;
  border-width: 1px;
  padding: calc(8px - 0px) calc(24px - 0px);
}

.circuit-7[href] {
  display: inline-block;
}

.circuit-7:disabled,
.circuit-7[disabled] {
  opacity: 0.4;
  pointer-events: none;
  -webkit-user-selectable: none;
  -moz-user-selectable: none;
  -ms-user-selectable: none;
  user-selectable: none;
}

.circuit-2 {
  height: 24px;
  width: 24px;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.circuit-0 {
  width: 24px;
  height: 24px;
  border-radius: 100%;
  border: 2px solid #FFFFFF;
  border-top-color: transparent;
  -webkit-animation: animation-0 1s infinite linear;
  animation: animation-0 1s infinite linear;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  opacity: 0;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
}

.circuit-4 {
  display: inline-block;
  opacity: 1;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  -webkit-transition: opacity 200ms ease-in-out,-webkit-transform 200ms ease-in-out;
  -webkit-transition: opacity 200ms ease-in-out,transform 200ms ease-in-out;
  transition: opacity 200ms ease-in-out,transform 200ms ease-in-out;
}

<button
  class="circuit-6 circuit-7 circuit-8"
>
  <div
    class="circuit-2 circuit-3"
  >
    <div
      aria-busy="true"
      aria-live="assertive"
      class="circuit-0 circuit-1"
      role="alert"
    />
  </div>
  <span
    class="circuit-4 circuit-5"
  />
</button>
`;

exports[`Button Style tests should have isLoading styles 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate3d(0,0,1,0deg);
    -ms-transform: rotate3d(0,0,1,0deg);
    transform: rotate3d(0,0,1,0deg);
  }

  100% {
    -webkit-transform: rotate3d(0,0,1,360deg);
    -ms-transform: rotate3d(0,0,1,360deg);
    transform: rotate3d(0,0,1,360deg);
  }
}

@keyframes animation-0 {
  0% {
    -webkit-transform: rotate3d(0,0,1,0deg);
    -ms-transform: rotate3d(0,0,1,0deg);
    transform: rotate3d(0,0,1,0deg);
  }

  100% {
    -webkit-transform: rotate3d(0,0,1,360deg);
    -ms-transform: rotate3d(0,0,1,360deg);
    transform: rotate3d(0,0,1,360deg);
  }
}

.circuit-2 {
  height: 24px;
  width: 24px;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.circuit-0 {
  width: 24px;
  height: 24px;
  border-radius: 100%;
  border: 2px solid #FFFFFF;
  border-top-color: transparent;
  -webkit-animation: animation-0 1s infinite linear;
  animation: animation-0 1s infinite linear;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  opacity: 0;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
}

.circuit-4 {
  display: inline-block;
  opacity: 1;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  -webkit-transition: opacity 200ms ease-in-out,-webkit-transform 200ms ease-in-out;
  -webkit-transition: opacity 200ms ease-in-out,transform 200ms ease-in-out;
  transition: opacity 200ms ease-in-out,transform 200ms ease-in-out;
}

.circuit-7 {
  background-color: #FAFBFC;
  border-color: #D8DDE1;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: inset 0 1px 0 1px rgba(255,255,255,0.06);
  display: block;
  color: #5C656F;
  cursor: pointer;
  font-weight: 700;
  width: auto;
  height: auto;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 16px;
  line-height: 24px;
  padding: calc(8px - 0px) calc(24px - 0px);
  overflow-y: hidden;
  pointer-events: none;
  position: relative;
  overflow: hidden;
}

.circuit-7:active {
  background-color: #D8DDE1;
  border-color: #9DA7B1;
  box-shadow: inset 0 4px 8px 0 rgba(12,15,20,0.3);
}

.circuit-7:focus {
  border-color: #9DA7B1;
  border-width: 2px;
  outline: 0;
  padding: calc(8px - 1px) calc(24px - 1px);
}

.circuit-7:hover {
  background-color: #D8DDE1;
}

.circuit-7:hover,
.circuit-7:active {
  border-color: #9DA7B1;
  border-width: 1px;
  padding: calc(8px - 0px) calc(24px - 0px);
}

.circuit-7[href] {
  display: inline-block;
}

.circuit-7:disabled,
.circuit-7[disabled] {
  opacity: 0.4;
  pointer-events: none;
  -webkit-user-selectable: none;
  -moz-user-selectable: none;
  -ms-user-selectable: none;
  user-selectable: none;
}

<button
  class="circuit-6 circuit-7 circuit-8"
>
  <div
    class="circuit-2 circuit-3"
  >
    <div
      aria-busy="true"
      aria-live="assertive"
      class="circuit-0 circuit-1"
      role="alert"
    />
  </div>
  <span
    class="circuit-4 circuit-5"
  />
</button>
`;
